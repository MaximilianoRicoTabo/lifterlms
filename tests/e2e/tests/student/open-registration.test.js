/**
 * Test the Setup Wizard
 *
 * @since 3.37.8
 * @since 3.37.14 Fix package references.
 * @since 4.5.0 Use package functions.
 * @since 4.12.0 Added registration test with a voucher.
 * @since [version] Added tests for form field localization (country, state, etc...).
 */

import {
	clickAndWait,
	createVoucher,
	fillField,
	logoutUser,
	registerStudent,
	select2Select,
	toggleOpenRegistration,
	visitPage,
} from '@lifterlms/llms-e2e-test-utils';

import { visitAdminPage } from '@wordpress/e2e-test-utils';

let openRegStatus = null;

/**
 * Toggles the open registration setting on or off
 *
 * @since 3.37.8
 * @since 4.5.0 Use toggleOpenRegistration function from utils pacakage.
 *
 * @param  {Boolean} status Whether to toggle on (`true`) or off (`false`).
 * @return {void}
 */
const toggleOpenReg = async function( status ) {

	if ( openRegStatus === status ) {
		return;
	}
	await toggleOpenRegistration( status );

}

describe( 'OpenRegistration', () => {

	afterEach( async () => {
		await logoutUser();
	} );

	describe( 'Registration', () => {

		it ( 'should not allow registration because user is already logged in.', async () => {

			await toggleOpenReg( true );
			await visitPage( 'dashboard' );
			await expect( await page.$( '.llms-new-person-form-wrapper > h4.llms-form-heading' ) ).toBeNull();

		} );

		it ( 'should allow registration.', async () => {

			await toggleOpenReg( true );
			await logoutUser();
			await visitPage( 'dashboard' );
			expect( await page.$eval( '.llms-new-person-form-wrapper > h4.llms-form-heading', el => el.textContent ) ).toBe( 'Register' );

		} );

		it ( 'should register a new user.', async () => {

			await toggleOpenReg( true );
			await registerStudent();
			expect( await page.$eval( 'h2.llms-sd-title', el => el.textContent ) ).toBe( 'Dashboard' );

		} );

		it ( 'should register a new user with a voucher.', async() => {

			await toggleOpenReg( true );
			const codes = await createVoucher( { codes: 1, uses: 1 } );
			await registerStudent( { voucher: codes[0] } );

			expect( await page.$eval( 'h2.llms-sd-title', el => el.textContent ) ).toBe( 'Dashboard' );

			await page.waitForSelector( '.llms-notification .llms-notification-main' );
			expect( await page.$eval( '.llms-notification .llms-notification-main', el => el.innerHTML ) ).toMatchSnapshot();

		} );

		it ( 'should not allow registration because open registration is disabled.', async () => {

			await toggleOpenReg( false );
			await logoutUser();
			await visitPage( 'dashboard' );
			await expect( await page.$( '.llms-new-person-form-wrapper > h4.llms-form-heading' ) ).toBeNull();

		} );

	} );

	describe( 'Localization', () => {

		it ( 'should localize city, state, and postcode fields when changing the selected country', async () => {

			await toggleOpenReg( true );
			await logoutUser();
			await visitPage( 'dashboard' );

			// China.
			await select2Select( '#llms_billing_country', 'China' );
			expect( await page.$eval( '.llms-l10n-state-select', el => el.innerHTML ) ).toMatchSnapshot();
			expect( await page.$eval( 'label[for="llms_billing_city"]', el => el.textContent ) ).toMatchSnapshot();
			expect( await page.$eval( 'label[for="llms_billing_zip"]', el => el.textContent ) ).toMatchSnapshot();

			// Peru changes name of the State & City fields.
			await select2Select( '#llms_billing_country', 'Peru' );
			expect( await page.$eval( '.llms-l10n-state-select', el => el.innerHTML ) ).toMatchSnapshot();
			expect( await page.$eval( 'label[for="llms_billing_city"]', el => el.textContent ) ).toMatchSnapshot();
			expect( await page.$eval( 'label[for="llms_billing_zip"]', el => el.textContent ) ).toMatchSnapshot();

			// United States.
			await select2Select( '#llms_billing_country', 'United States' );
			expect( await page.$eval( '.llms-l10n-state-select', el => el.innerHTML ) ).toMatchSnapshot();
			expect( await page.$eval( 'label[for="llms_billing_city"]', el => el.textContent ) ).toMatchSnapshot();
			expect( await page.$eval( 'label[for="llms_billing_zip"]', el => el.textContent ) ).toMatchSnapshot();

			// UAB has no postal code or city.
			await select2Select( '#llms_billing_country', 'United Arab Emirates' );
			expect( await page.$eval( '.llms-l10n-state-select', el => el.innerHTML ) ).toMatchSnapshot();
			expect( await page.$eval( '#llms_billing_city', el => el.disabled ) ).toBe( true );
			expect( await page.$eval( '#llms_billing_zip', el => el.disabled ) ).toBe( true );

			// Tokelau has no states.
			await select2Select( '#llms_billing_country', 'Tokelau' );
			expect( await page.$eval( '#llms_billing_state', el => el.disabled ) ).toBe( true );
			expect( await page.$eval( 'label[for="llms_billing_city"]', el => el.textContent ) ).toMatchSnapshot();
			expect( await page.$eval( 'label[for="llms_billing_zip"]', el => el.textContent ) ).toMatchSnapshot();

		} );

	} );

} );
