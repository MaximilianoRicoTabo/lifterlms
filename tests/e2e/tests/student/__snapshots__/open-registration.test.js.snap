// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 1`] = `"Province*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 2`] = `"City*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 3`] = `"Postal code*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 4`] = `"Region*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 5`] = `"District*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 6`] = `"Postal code*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 7`] = `"State*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 8`] = `"City*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 9`] = `"ZIP code*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 10`] = `"Emirate*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 11`] = `"City*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country 12`] = `"Postal code*"`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country: China 1`] = `
Object {
  "AH": "Anhui",
  "BJ": "Beijing",
  "CQ": "Chongqing",
  "FJ": "Fujian",
  "GD": "Guangdong",
  "GS": "Gansu",
  "GX": "Guangxi Zhuang Autonomous Region",
  "GZ": "Guizhou",
  "HA": "Henan",
  "HB": "Hubei",
  "HE": "Hebei",
  "HI": "Hainan",
  "HK": "Hong Kong",
  "HL": "Heilongjiang",
  "HN": "Hunan",
  "JL": "Jilin",
  "JS": "Jiangsu",
  "JX": "Jiangxi",
  "LN": "Liaoning",
  "MO": "Macau",
  "NM": "Inner Mongolia",
  "NX": "Ningxia Hui Autonomous Region",
  "QH": "Qinghai",
  "SC": "Sichuan",
  "SD": "Shandong",
  "SH": "Shanghai",
  "SN": "Shaanxi",
  "SX": "Shanxi",
  "TW": "Taiwan Province, People's Republic of China",
  "TW-KEE": "Keelung",
  "XJ": "Xinjiang",
  "XZ": "Tibet Autonomous Region",
  "YN": "Yunnan",
  "ZJ": "Zhejiang",
}
`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country: Peru 1`] = `
Object {
  "AMA": "Amazonas",
  "ANC": "Áncash",
  "APU": "Apurímac",
  "ARE": "Arequipa",
  "AYA": "Ayacucho",
  "CAJ": "Cajamarca",
  "CAL": "Callao",
  "CUS": "Cusco",
  "HUC": "Huanuco",
  "HUV": "Huancavelica",
  "ICA": "Ica",
  "JUN": "Junín",
  "LAL": "La Libertad",
  "LAM": "Lambayeque",
  "LIM": "Lima",
  "MDD": "Madre de Dios",
  "MOQ": "Moquegua",
  "PAS": "Pasco",
  "PIU": "Piura",
  "PUN": "Puno",
  "SAM": "San Martín",
  "TAC": "Tacna",
  "TUM": "Tumbes",
  "UCA": "Ucayali",
}
`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country: Tokelau 1`] = `
Object {
  " ": " ",
}
`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country: United Arab Emirates 1`] = `
Object {
  "AJ": "Ajman Emirate",
  "AZ": "Abu Dhabi Emirate",
  "DU": "Dubai",
  "FU": "Fujairah",
  "RK": "Ras al-Khaimah",
  "SH": "Sharjah Emirate",
  "UQ": "Umm al-Quwain",
}
`;

exports[`OpenRegistration Localization should localize city, state, and postcode fields when changing the selected country: United States 1`] = `
Object {
  "AK": "Alaska",
  "AL": "Alabama",
  "AR": "Arkansas",
  "AS": "American Samoa",
  "AZ": "Arizona",
  "CA": "California",
  "CO": "Colorado",
  "CT": "Connecticut",
  "DC": "District of Columbia",
  "DE": "Delaware",
  "FL": "Florida",
  "GA": "Georgia",
  "GU": "Guam",
  "HI": "Hawaii",
  "IA": "Iowa",
  "ID": "Idaho",
  "IL": "Illinois",
  "IN": "Indiana",
  "KS": "Kansas",
  "KY": "Kentucky",
  "LA": "Louisiana",
  "MA": "Massachusetts",
  "MD": "Maryland",
  "ME": "Maine",
  "MI": "Michigan",
  "MN": "Minnesota",
  "MO": "Missouri",
  "MP": "Northern Mariana Islands",
  "MS": "Mississippi",
  "MT": "Montana",
  "NC": "North Carolina",
  "ND": "North Dakota",
  "NE": "Nebraska",
  "NH": "New Hampshire",
  "NJ": "New Jersey",
  "NM": "New Mexico",
  "NV": "Nevada",
  "NY": "New York",
  "OH": "Ohio",
  "OK": "Oklahoma",
  "OR": "Oregon",
  "PA": "Pennsylvania",
  "PR": "Puerto Rico",
  "RI": "Rhode Island",
  "SC": "South Carolina",
  "SD": "South Dakota",
  "TN": "Tennessee",
  "TX": "Texas",
  "UM": "United States Minor Outlying Islands",
  "UT": "Utah",
  "VA": "Virginia",
  "VI": "United States Virgin Islands",
  "VT": "Vermont",
  "WA": "Washington",
  "WI": "Wisconsin",
  "WV": "West Virginia",
  "WY": "Wyoming",
}
`;

exports[`OpenRegistration Registration should register a new user with a voucher. 1`] = `" <h4 class=\\"llms-notification-title\\">You've earned a certificate!</h4> <div class=\\"llms-notification-body\\"><div class=\\"llms-mini-cert\\"> <h2 class=\\"llms-mini-cert-title\\">Awarded on Registration</h2> <div class=\\"llms-mini-cert--body \\"> <div class=\\"llms-mini-cert--mock-line\\"></div> <div class=\\"llms-mini-cert--mock-line\\"></div> <div class=\\"llms-mini-cert--mock-line\\"></div> <div class=\\"llms-mini-cert--mock-line\\"></div> <div class=\\"llms-mini-cert--mock-line\\"></div> <div class=\\"llms-mini-cert--mock-dot\\"></div> <div class=\\"llms-mini-cert--mock-line\\"></div> <p></p></div> <p></p></div> </div> "`;
